File: mvvm/src/main/java/com/rui/mvvm/di/scopes/ViewModelScope.java
package com.rui.mvvm.di.scopes;


import androidx.lifecycle.ViewModel;

import dagger.MapKey;

import java.lang.annotation.*;

/**
 * Created by rui on 2018/10/29.
 * Dagger ViewModelScope
 */
@Documented
@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@MapKey
public @interface ViewModelScope {
    Class<? extends ViewModel> value();
}

File: mvvm/src/main/java/com/rui/mvvm/di/scopes/ViewModelScope.kt
package com.rui.mvvm.di.scopes

import androidx.lifecycle.ViewModel
import dagger.MapKey
import java.lang.annotation.Documented
import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy
import kotlin.reflect.KClass

/**
 * Created by rui on 2018/10/29.
 * Dagger ViewModelScope
 */
@Documented
@Target(
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER
)
@Retention(
    RetentionPolicy.RUNTIME
)
@MapKey
annotation class ViewModelScope(val value: KClass<out ViewModel>)